src/utils.ts

```diff
@@ -54,6 +43,123 @@ const selectNextRelay = async (
   return undefined;
 };
 
+const estimateInternalCallGas = async ({
+  internalEstimationCorrection,
+  estimatedGasCorrectionFactor,
+  ...estimateGasParams
+}: EstimateInternalGasParams): Promise<BigNumber> => {
+  const provider = getProvider();
+
+  let estimation: BigNumber = await provider.estimateGas(
+    estimateGasParams
+  );
+
+  estimation = applyInternalEstimationCorrection(
+    estimation,
+    internalEstimationCorrection
+  );
+
+  return applyGasCorrectionFactor(estimation, estimatedGasCorrectionFactor);
+}
+
+const estimateTokenTransferGas = async ({
+  internalEstimationCorrection,
+  estimatedGasCorrectionFactor,
+  preDeploySWAddress,
+  relayRequest,
+}: TokenGasEstimationParams): Promise<BigNumber> => {
+  const {
+    request: { tokenContract, tokenAmount },
+    relayData: { callForwarder, gasPrice, feesReceiver },
+  } = relayRequest;
+
+  if (!Number(tokenContract) || tokenAmount.toString() === '0') {
+    return constants.Zero;
+  }
+
+  let tokenOrigin: string | undefined;
+
+  if (isDeployRequest(relayRequest)) {
+    tokenOrigin = preDeploySWAddress;
+
+    // If it is a deploy and tokenGas was not defined, then the smartwallet address
+    // is required to estimate the token gas. This value should be calculated prior to
+    // the call to this function
+    if (!tokenOrigin || tokenOrigin === constants.AddressZero) {
+      throw Error(MISSING_SMART_WALLET_ADDRESS);
+    }
+  } else {
+    tokenOrigin = callForwarder.toString();
+
+    if (tokenOrigin === constants.AddressZero) {
+      throw Error(MISSING_CALL_FORWARDER);
+    }
+  }
+
+  const provider = getProvider();
+
+  const erc20 = IERC20__factory.connect(
+    tokenContract.toString(),
+    provider
+  );
+  const gasCost = await erc20.estimateGas.transfer(
+    feesReceiver,
+    tokenAmount,
+    { from: tokenOrigin, gasPrice }
+  );
+
+  const internalCallCost = applyInternalEstimationCorrection(
+    gasCost,
+    internalEstimationCorrection
+  );
+
+  return applyGasCorrectionFactor(
+    internalCallCost,
+    estimatedGasCorrectionFactor
+  );
+}
+
+const getSmartWalletAddress = async (
```

Comment:
In the future, we should consider if we want to replace this mechanism with the one that doesn't use the factory to check if the result is the same and if we can avoid using the factory.

Reply:
updated